### PR TITLE
(maint) HTTP docs

### PR DESCRIPTION
**Pull request by joshcooper**
Proposal for HTTP client in puppet

### CODE REVIEW

File: docs/http.md [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+* DNS SRV
+* Server list
+* Default puppet settings
+
+If the `route_to` method yields a route, but it results in an exception, such as
+"connection refused", then the session will yield the next route, until all
+routes for all resolvers have been tried.
+
+If the caller successfully uses a route, then the session will yield the same
+route the next time `route_to` is called again.
```
Comment (kris-bosland) [inline docs/http.md]: I assume you mean, that route will be the first tried, before following the priority list if that route cannot be reached?

File: docs/http.md [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+Puppet agents support downloading file content from 3rd party file servers,
+which reduces load on the compiler. The `Client` will provide a low-level API
+for making `GET` requests for an arbitrary URL, and streaming the response body.
+
+Puppet only trusts the puppet PKI for its REST requests. However, it should be
+possible to additionally trust the system store when making HTTPS requests:
+
+```ruby
+client = Puppet::HTTP::Client.new
+response = http.get_content("https://artifactory.example.com/java.tar.gz", :trust_system_store: true)
```
Comment (kris-bosland) [inline docs/http.md]: It feels like there will be common modes for trust_system_store: hard true/false on the call, and sometimes depending on a setting.  Not sure if we should get into those details here.

File: docs/http.md [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+  end
+
+  # etc
+end
+```
+
+And register it with puppet:
+
+```ruby
+Puppet.push_context(http_client: HttpClientAdapter.new(Puppet::Server::HttpClient.new))
```
Comment (kris-bosland) [inline docs/http.md]: 👍

File: docs/http.md (hunk 2 above)
Comment (joshcooper) [reply docs/http.md]: For puppet REST calls, we want to only trust the puppet PKI, for example, so we don't accidentally download a malicious catalog. There only case I know for trusting the system store are when downloading file content, or when the ["http" report processor](https://github.com/puppetlabs/puppet/blob/6.7.0/lib/puppet/reports/http.rb#L32). The former is something that will always be enabled. The former will likely be a puppet setting (https://tickets.puppetlabs.com/browse/PUP-7737).

File: docs/http.md [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
```
[the remark this answers is not in the document]
Comment (joshcooper) [reply docs/http.md]: Actually I take that back. If puppet server installs/registers its own client, then there wouldn't be pool.

File: docs/http.md [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+
+#### Client
+
+Has a pool of persistent HTTP connections and methods for creating an
+HTTP session. Closes persistent connections when its close method is called.
+
+Has low-level HTTP methods, such as `get`, `post`, etc which take the path,
+headers, options, and allow the caller to stream the request and response body.
+
+#### Connection Pool
```
[the remark this answers is not in the document]
Comment (joshcooper) [reply docs/http.md]: Mentioned above that the connection pool wouldn't be used for puppetserver

**Comment by joshcooper** [comment]
I change `Route` to be the static definition that describes the web service, and `Service` to be the server/endpoint/api that the client is talking to. The `Service` has methods that are appropriate, eg `Puppet::HTTP::Service::Ca#get_certificate(name)`.

**Pull request merged**